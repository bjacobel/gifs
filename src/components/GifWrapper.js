import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

import Gif from './Gif';
import {
  rootURL,
  thumbURL
} from '../constants';
import * as clipboard from '../services/clipboard';
import { watchForSize } from '../actions/gifs';
import {
  animateGif,
  freezeGif,
  saveMostRecentAnimation
} from '../actions/animation';

function mapStateToProps(state) {
  return {
    animation: state.animation
  };
}

const mapDispatchToProps = {
  animateGif,
  freezeGif,
  saveMostRecentAnimation,
  watchForSize
};

export default class GifWrapper extends Component {
  shouldComponentUpdate(nextProps) {
    // Only update if animation[this.id] changes

    const { animation, gif } = this.props;

    if (nextProps.animation.hasOwnProperty([gif.id])) {
      if (animation.hasOwnProperty([gif.id])) {
        return animation[gif.id] !== nextProps.animation[gif.id];
      }
      return true;
    }

    return false;
  }

  render() {
    const {
      gif,
      animation,
      animateGif,
      freezeGif,
      saveMostRecentAnimation,
      watchForSize
    } = this.props;

    const enabled = animation[gif.id] || false;
    const gifURL = enabled ? `${rootURL}${gif.src}` : `${thumbURL}${gif.src}`;
    const image = new Image();
    image.src = gifURL;

    const enableMotion = () => {
      animateGif(gif.id);
      saveMostRecentAnimation(gif.id);
    };
    const disableMotion = () => { freezeGif(gif.id); };
    const clip = () => { clipboard.copy(rootURL + gif.src); };
    const disableAndClip = () => { disableMotion(); clip(); };

    return (
      <div className="gif-wrapper"
        onMouseOver={ enableMotion }
        onMouseOut={ disableMotion }
        onTouchStart={ enableMotion }
        onTouchEnd={ disableAndClip }
        onMouseUp={ clip }
      >
        <Gif
          image={ image }
          watchForSize={ watchForSize }
          id={ gif.id }
          name={ gif.src }
        />
      </div>
    );
  }
}

GifWrapper.propTypes = {
  gif: PropTypes.shape({
    id: PropTypes.string.isRequired,
    src: PropTypes.string.isRequired
  }).isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GifWrapper);
