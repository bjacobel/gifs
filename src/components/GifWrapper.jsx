import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import StaticGif from './StaticGif';
import AnimatedGif from './AnimatedGif';
import GifTags from './GifTags';
import { rootURL } from '../constants';
import * as clipboard from '../services/clipboard';
import { animateGif, freezeGif } from '../actions/animation';

function mapStateToProps(state) {
  return {
    animation: state.animation
  };
}

const mapDispatchToProps = {
  animateGif,
  freezeGif
};

export default class GifWrapper extends Component {
  shouldComponentUpdate(nextProps) {
    // Only update if animation[this.id] changes

    const { animation, gif } = this.props;

    if (nextProps.animation.hasOwnProperty([gif.id])) {
      if (animation.hasOwnProperty([gif.id])) {
        return animation[gif.id] !== nextProps.animation[gif.id];
      }
      return true;
    }

    return false;
  }

  render() {
    const {
      gif,
      tags,
      animation,
      animateGif, // eslint-disable-line no-shadow
      freezeGif // eslint-disable-line no-shadow
    } = this.props;

    const enabled = animation[gif.id] || false;

    const img = new Image();
    img.src = rootURL + gif.src;

    const enableMotion = () => { animateGif(gif.id); };
    const disableMotion = () => { freezeGif(gif.id); };
    const clip = () => { clipboard.copy(rootURL + gif.src); };
    const disableAndClip = () => { disableMotion(); clip(); };

    return (
      <div className="gif-wrapper">
        <div
          onMouseOver={ enableMotion }
          onMouseOut={ disableMotion }
          onTouchStart={ enableMotion }
          onTouchEnd={ disableAndClip }
          onMouseUp={ clip }
          className={ classNames('swapper', { enabled }) }
        >
          <AnimatedGif img={ img }/>
          <StaticGif img={ img } id={ gif.id }/>
        </div>
        <GifTags tags={ tags }/>
      </div>
    );
  }
}

GifWrapper.propTypes = {
  gif: PropTypes.shape({
    id: PropTypes.string.isRequired,
    src: PropTypes.string.isRequired
  }).isRequired,
  tags: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired
    }).isRequired
  ).isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GifWrapper);
