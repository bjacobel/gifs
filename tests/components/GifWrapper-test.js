import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import GifWrapper from '../../src/components/GifWrapper';

jest.mock('../../src/services/image');

describe('GifWrapper component', () => {
  it('matches the snapshot', () => {
    const enzymeRepr = shallow(
      <GifWrapper.WrappedComponent
        gif={ { id: '1', src: '//:0' } }
        animation={ { 1: false } }
        animateGif={ jest.fn() }
        freezeGif={ jest.fn() }
        saveMostRecentAnimation={ jest.fn() }
        watchForSize={ jest.fn() }
      />,
    );

    expect(enzymeRepr).toMatchSnapshot();
  });

  describe('pointer events', () => {
    it('calls enableMotion on mouseOver', () => {
      const animate = jest.fn();
      const saveAnimation = jest.fn();

      const gifWrapper = shallow(
        <GifWrapper.WrappedComponent
          gif={ { id: '1', src: '//:0' } }
          animation={ { 1: true } }
          animateGif={ animate }
          freezeGif={ jest.fn() }
          saveMostRecentAnimation={ saveAnimation }
          watchForSize={ jest.fn() }
        />,
      );

      gifWrapper.simulate('mouseOver');

      expect(animate).toHaveBeenCalled();
      expect(saveAnimation).toHaveBeenCalled();
    });

    it('calls disableMotion on mouseOut', () => {
      const freezeGif = jest.fn();

      const gifWrapper = shallow(
        <GifWrapper.WrappedComponent
          gif={ { id: '1', src: '//:0' } }
          animation={ { 1: false } }
          animateGif={ jest.fn() }
          freezeGif={ freezeGif }
          saveMostRecentAnimation={ jest.fn() }
          watchForSize={ jest.fn() }
        />,
      );

      gifWrapper.simulate('mouseOut');
      expect(freezeGif).lastCalledWith('1');
    });

    it('calls toggleMotion on each click', () => {
      const animateGif = jest.fn();
      const freezeGif = jest.fn();

      const gifWrapper = shallow(
        <GifWrapper.WrappedComponent
          gif={ { id: '1', src: '//:0' } }
          animation={ { 1: false } }
          animateGif={ animateGif }
          freezeGif={ freezeGif }
          saveMostRecentAnimation={ jest.fn() }
          watchForSize={ jest.fn() }
        />,
      );

      gifWrapper.simulate('touchEnd');
      expect(animateGif).toHaveBeenCalled();

      gifWrapper.setProps({ animation: { 1: true } });

      gifWrapper.simulate('touchEnd');
      expect(freezeGif).toHaveBeenCalled();
    });
  });

  describe('shouldComponentUpdate', () => {
    let component;
    let renderSpy;
    let sandbox;

    beforeEach(() => {
      sandbox = sinon.sandbox.create();

      component = (
        <GifWrapper.WrappedComponent
          gif={ { id: '1', src: '//:0' } }
          animation={ { 1: false } }
          animateGif={ jest.fn() }
          freezeGif={ jest.fn() }
          saveMostRecentAnimation={ jest.fn() }
          watchForSize={ jest.fn() }
        />
      );

      renderSpy = sandbox.spy(GifWrapper.WrappedComponent.prototype, 'render');
    });

    afterEach(() => {
      sandbox.restore();
    });

    it("doesn't update unless animation[that gif's id] changes", () => {
      const enzymeRepr = shallow(component);
      expect(renderSpy.calledOnce).toBeTruthy();
      enzymeRepr.setProps({ animation: { 1: false } });
      expect(renderSpy.calledOnce).toBeTruthy();
    });

    it('does update when animation[id for that gif] changes', () => {
      const enzymeRepr = shallow(component);
      expect(renderSpy.calledOnce).toBeTruthy();
      enzymeRepr.setProps({ animation: { 1: true } });
      expect(renderSpy.calledTwice).toBeTruthy();
    });

    it("doesn't update when nextProps animation doesn't include the id for the gif", () => {
      const enzymeRepr = shallow(component);
      expect(renderSpy.calledOnce).toBeTruthy();
      enzymeRepr.setProps({ animation: { 2: false } });
      expect(renderSpy.calledOnce).toBeTruthy();
    });

    it("updates when original animation doesn't include the id for the gif, but nextProps does", () => {
      const enzymeRepr = shallow(
        <GifWrapper.WrappedComponent
          gif={ { id: '1', src: '//:0' } }
          animation={ { 2: false } }
          animateGif={ jest.fn() }
          freezeGif={ jest.fn() }
          saveMostRecentAnimation={ jest.fn() }
          watchForSize={ jest.fn() }
        />,
      );
      expect(renderSpy.calledOnce).toBeTruthy();
      enzymeRepr.setProps({ animation: { 1: true } });
      expect(renderSpy.calledTwice).toBeTruthy();
    });
  });
});
