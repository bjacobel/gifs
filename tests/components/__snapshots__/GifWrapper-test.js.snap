exports[`GifWrapper component matches the snapshot 1`] = `
<div
  className="gif-wrapper"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onTouchEnd={[Function]}>
  <Gif
    id="1"
    image={
      HTMLImageElement {
        "addEventListener": [Function],
        "appendChild": [Function],
        "blur": [Function],
        "click": [Function],
        "cloneNode": [Function],
        "compareDocumentPosition": [Function],
        "constructor": [Function],
        "contains": [Function],
        "dispatchEvent": [Function],
        "focus": [Function],
        "getAttribute": [Function],
        "getAttributeNS": [Function],
        "getAttributeNames": [Function],
        "getAttributeNode": [Function],
        "getAttributeNodeNS": [Function],
        "getBoundingClientRect": [Function],
        "getClientRects": [Function],
        "getElementsByClassName": [Function],
        "getElementsByTagName": [Function],
        "getElementsByTagNameNS": [Function],
        "hasAttribute": [Function],
        "hasAttributeNS": [Function],
        "hasAttributes": [Function],
        "hasChildNodes": [Function],
        "insertAdjacentHTML": [Function],
        "insertBefore": [Function],
        "isEqualNode": [Function],
        "matches": [Function],
        "normalize": [Function],
        "querySelector": [Function],
        "querySelectorAll": [Function],
        "remove": [Function],
        "removeAttribute": [Function],
        "removeAttributeNS": [Function],
        "removeAttributeNode": [Function],
        "removeChild": [Function],
        "removeEventListener": [Function],
        "replaceChild": [Function],
        "setAttribute": [Function],
        "setAttributeNS": [Function],
        "setAttributeNode": [Function],
        "setAttributeNodeNS": [Function],
        "src": "https://gifthumbs.bjacobel.com///:0",
        "toString": [Function],
        "webkitMatchesSelector": [Function],
      }
    }
    name="//:0"
    watchForSize={[Function]} />
</div>
`;
